### PR TITLE
fixing logger, adding default to no color given not the terminal

### PR DESCRIPTION
**Pull request by vsoch**
Signed-off-by: Vanessa Sochat <vsochat@stanford.edu>

**Description of the Pull Request (PR):**

This pull request will detect if the user is not directing output to the terminal, and if not, disable color automatically. It also fixes a bug that @bauerm97 and I overlooked of missing a few of the message levels! Specifically this part:

```go
// DisableColor for the logger
func DisableColor() {
	messageColors = map[messageLevel]string{
		fatal:    "",
		error:    "",
		warn:     "",
		info:     "",
	}
	colorReset = ""
}
```

was missing levels!

```go
// DisableColor for the logger
func DisableColor() {
	messageColors = map[messageLevel]string{
		fatal:    "",
		error:    "",
		warn:     "",
		info:     "",
		verbose:  "",
		verbose2: "",
		verbose3: "",
		debug:    "",
	}
	colorReset = ""
}
```

Thus the debug still had weird characters with the flag.

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by bauerm97** [comment]
Hey @vsoch, maybe since we already identified this as potentially buggy (we both missed a bug already at first pass) there should be some unit tests to ensure that we don't have any more regressions. A simple `sylog_test.go` file could be added which does a few tests. Would you be willing to take a stab? 

A simple baseline test that does the following would suffice:
1. Call `prefix()` against values of `level` (input to the function) and `loggerLevel` (level of logger verbosity, use `SetLevel()` to change this). This is effectively an N^2 matrix of variables
2. Check returned prefixes contain the appropriate color strings
3. Call `DisableColor()`
4. Run `prefix()` against the matrix again
5. Check that all returned prefixes do *not* contain color strings this time

**Comment by vsoch** [comment]
Good idea! I'm thinking also about the functionality - if we disable color automatically given output isn't to a terminal, do we need the --nocolor option? I think so, because there are terminals that don't support color where it would come out looking funky. On the flip side, what if someone is piping to a file and does want color? I can imagine someone wanting to pipe to a file that is printed later. In this case, given that we have a --nocolor argument, I think there should be a --color flag. Or we can consolidate those two into one thing and have something like `--color` or `--color=off` (does GoLang support having a flag that can serve as a boolean and a string argument?) We could also completely remove --nocolor, assuming that the default of disabling when we pipe to a file will fit most use cases. Thoughts?

**Comment by vsoch** [comment]
@bauerm97 how do I call private functions? I tried adding -v and capitalizing prefix, for example, neither of those work.

**Comment by vsoch** [comment]
I'm having quite a bit of trouble writing tests for private functions (it seems like it's not common to do) but I updated the PR with the bug fixes.

**Comment by bauerm97** [comment]
@vsoch actually I think it's fairly common. I'm pretty sure you simply declare your test file as part of the same package. In `sylog_test.go`:

```
package sylog

...
```

Also you do need to ensure that you're using the `sylog` build tag, otherwise it will compile with the `sylog_dummy.go` file which will not have a `prefix()` function. I would hazard a guess that this is what's causing issues

**Comment by vsoch** [comment]
I had the package sylog at the top, what is this build tag that you speak of? I saw this header in the sylog.go file:

```
// +build sylog
```
And I added it to my test file, but I got the same error.. Here is everything (a bit messy atm) that I have if you want to take a look. I'm running with `go test -v .`

```go
// Copyright (c) 2019, Sylabs Inc. All rights reserved.
// This software is licensed under a 3-clause BSD license. Please consult the
// LICENSE.md file distributed with the sources of this project regarding your
// rights to use or distribute this software.

// +build sylog

package sylog

import (
        "fmt"
	"testing"

	"github.com/magiconair/properties/assert"
)

var nocolor bool
type messageLevel int

const (
	fatal    messageLevel = iota - 4 // fatal    : -4
	error                            // error    : -3
	warn                             // warn     : -2
	log                              // log      : -1
	_                                // SKIP     : 0
	info                             // info     : 1
	verbose                          // verbose  : 2
	verbose2                         // verbose2 : 3
	verbose3                         // verbose3 : 4
	debug                            // debug    : 5
)

var messageColors = map[messageLevel]string{
	fatal:    "\x1b[31m",
	error:    "\x1b[31m",
	warn:     "\x1b[33m",
	info:     "\x1b[34m",
	log:      "\x1b[0m",
	verbose:  "\x1b[0m",
	verbose2: "\x1b[0m",
	verbose3: "\x1b[0m",
	debug:    "\x1b[0m",
}

func TestSetLevel(t *testing.T) {
        var loggerLevel int

        // Loop through logging levels and check that they are set
        for level := -4; level <= 5; level++ {
                SetLevel(level)
                loggerLevel = GetLevel()
                assert.Equal(t, loggerLevel, level, "The level should be set.")
        }
}


//func TestSylogDisableColor(m *testing.T) {
//		sylog.DisableColor()
//}
func TestSylogPrefix(m *testing.T) {
	var messagePrefix string
        type level int

        // Call prefix() against values of level
        for level := -4; level <= 5; level++ {
                SetLevel(level)
                messagePrefix = prefix(level)
                fmt.Println(messagePrefix)                
                // Check returned prefixes contain the appropriate color strings
        	//assert.T(prefix, level, "true", "The level should be set.")
        }



//Call DisableColor()
//Run prefix() against the matrix again
//Check that all returned prefixes do not contain color strings this time
}
```
The functions don't seem to be defined:

```
$ go test -v sylog_test.go 
# command-line-arguments [command-line-arguments.test]
./sylog_test.go:50:17: undefined: SetLevel
./sylog_test.go:51:31: undefined: GetLevel
./sylog_test.go:66:17: undefined: SetLevel
./sylog_test.go:67:33: undefined: prefix
FAIL	command-line-arguments [build failed]
```
Then if I just use a . (I've had better luck with this in the past):

```
$ go test -v .
?   	github.com/sylabs/singularity/internal/pkg/sylog	[no test files]
```

**Comment by bauerm97** [comment]
Hey @vsoch , sorry for the confusion here (wrote that post on my phone on a plane). What you need to do is run the `go test` command with the appropriate `sylog` tag set. To do this, run the command:

`go test -tags=sylog .`

By adding `-tags=sylog`, you tell Go to compile the test code with the `sylog` tag. This causes the proper `sylog.go` file to get included in the build.

**Comment by vsoch** [comment]
thanks @bauerm97! I'll try that.

**Comment by vsoch** [comment]
Hey @bauerm97 can you give me some kind of template to start with? I'm not making good use of time because I don't have a sense of how to do this. I think it will be better for me to learn from your example.

**Comment by vsoch** [comment]
This looks like a Travis error:

```bash
ERROR:   unpackSIF Failed: While copying partition data to bundle: cp Failed: exit status 1: cp: reading `/tmp/tmpmnt-610282689/./.singularity.d/Singularity': Input/output error
                cp: reading `/tmp/tmpmnt-610282689/./.singularity.d/actions/exec': Input/output error
...
                FATAL:   While performing build: packer failed to pack: While copying partition data to bundle: cp Failed: exit status 1: cp: reading `/tmp/tmpmnt-610282689/./.singularity.d/Singularity': Input/output error
```

Have you guys seen this before? Rerun?

File: cmd/internal/cli/singularity.go
Comment (bauerm97) [inline cmd/internal/cli/singularity.go:109]: I don't think we should automatically be trying to do this here. There are too many variables at play for that to make sense in my opinion. I think just providing the flag is good enough

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [inline internal/pkg/sylog/sylog_test.go]: None of these function names should be including `Sylog`. All of these functions are already within `package sylog`, so it's redundant

File: CHANGELOG.md
Comment (bauerm97) [inline CHANGELOG.md]: I think this should be removed from here

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [inline internal/pkg/sylog/sylog_test.go:35]: This line is unnecessary, we can just do `if GetLevel() == tt.level`

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [inline internal/pkg/sylog/sylog_test.go]: Delete empty line

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [inline internal/pkg/sylog/sylog_test.go]: Delete empty line

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [inline internal/pkg/sylog/sylog_test.go]: This comment is probably a copy paste error? Test that the prefix isn't colored, right?

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [inline internal/pkg/sylog/sylog_test.go]: Here, the suffix should _not_ contain a color reset string.

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [inline internal/pkg/sylog/sylog_test.go:113]: I'm not sure why this isn't failing, if we have called `DisableColor`, then we should never be including any color codes in our return value from `prefix()`

**Review by bauerm97 — CHANGES_REQUESTED** [review]
There are a few changes we still need to make to the testing package. Also an open question about this approach to testing: If we call `DisableColor` in one `TestXXX` function, are the subsequent tests running in that same context or is a new Go runtime spawned up. It might be possible that running these tests will result in different behavior depending on which test was ran first. 

cc @cclerget @tri-adam if you guys know the answer to that ^

File: cmd/internal/cli/singularity.go
Comment (vsoch) [reply cmd/internal/cli/singularity.go:109]: This was a suggestion / request of someone else (I can't remember now, @cclerget was it you?) that I added after the fact. It the maintainers want to decide on a consensus I'd be happy to update the functionality.

File: internal/pkg/sylog/sylog_test.go
Comment (vsoch) [reply internal/pkg/sylog/sylog_test.go]: Affirmative.

File: internal/pkg/sylog/sylog_test.go
Comment (vsoch) [reply internal/pkg/sylog/sylog_test.go:113]: That means it's correct - the error is only called if the return value (levelPrefix) has the suffix (which it doesn't because it's disabled).

File: internal/pkg/sylog/sylog_test.go
Comment (vsoch) [reply internal/pkg/sylog/sylog_test.go:35]: The reason is because we want to save the variable and report the result back to the user - we use it twice.

File: cmd/internal/cli/singularity.go
Comment (mem) [reply cmd/internal/cli/singularity.go:109]: I actually think this is nice to make. If we want color for example when using a pager (e.g. "singularity ... | less") then we need an additional option `--color` or follow git's lead (and other tools' as well) and have `--color=on`, `--color=off`, `--color=auto`, with the last one being the default value. `--color=auto` in git and other tools means "color when the output is a terminal, no color otherwise".

File: internal/pkg/sylog/sylog_test.go
Comment (mem) [reply internal/pkg/sylog/sylog_test.go:113]: That makes sense. Why do you need the extra tt.level != debug, though?

File: cmd/internal/cli/singularity.go
Comment (vsoch) [reply cmd/internal/cli/singularity.go:109]: Agree! This is similar to what I was bringing up a few days ago :point_up:  https://github.com/sylabs/singularity/pull/2841#issuecomment-471022566

File: internal/pkg/sylog/sylog_test.go
Comment (vsoch) [reply internal/pkg/sylog/sylog_test.go:113]: Good question! When debug is the level, there is the added uid and process id to the output (something like `[uid=1000,pid=1234]`) and it's not colored, so it doesn't have the same suffix as the others.

File: cmd/internal/cli/singularity.go
Comment (bauerm97) [reply cmd/internal/cli/singularity.go:109]: Alright, so if this is a standard scheme for coloring output, then I'm happy to oblige here.

Comment (vsoch) [reply cmd/internal/cli/singularity.go:109]: @mem and @bauerm97 do you want to add the `--color` variable as @mem detailed above, or have two booleans `--nocolor` and `--color` in addition to the automatic detection of where is being written (which is overridden by specifying one of these flags).

File: internal/pkg/sylog/sylog_test.go
Comment (bauerm97) [reply internal/pkg/sylog/sylog_test.go:113]: So, I think it's actually probably easier to just match a regex here instead of trying to handle every individual case. We can create a color regex, and a no-color regex:

To match for colored prefixes: `^(\\x1b\[\d+m)(\w{1,7}\s*)(\\x1b\[0m)`
To match for without prefixes: `^(FATAL|ERROR|WARNING|LOG|INFO|VERBOSE|(DEBUG\s*\[.*\]))\s*$`

I think we can just apply those regexps on the returned prefixes. The first should only match for prefixes returned in `TestPrefix`, and the second one should match only for prefixes returned within `TestDisableColor`. I think this will greatly simplify the possibility that we've gotten something wrong in the testing matrix

Comment (bauerm97) [reply internal/pkg/sylog/sylog_test.go:113]: Actually that second one needs a small tweak, hold on

Comment (bauerm97) [reply internal/pkg/sylog/sylog_test.go:113]: OK, second regex is correct now I believe

Comment (vsoch) [reply internal/pkg/sylog/sylog_test.go:113]: A single regex would be a lot cleaner, but Isn't the idea of matching the messageLevels to strings to check for that? For example, we want to check that "info" (the integer / messageLevel) corresponds to the string "INFO." If we do one gigantic regular expression, we don't check for that.

File: cmd/internal/cli/singularity.go
Comment (vsoch) [reply cmd/internal/cli/singularity.go:109]: ping! What is the preference for the argument?

**Comment by GodloveD** [comment]
@mem @bauerm97 @vsoch where does this stand?

File: cmd/internal/cli/singularity.go
Comment (vsoch) [reply cmd/internal/cli/singularity.go:109]: @GodloveD I was waiting for feedback here.

Comment (WestleyK) [reply cmd/internal/cli/singularity.go:109]: @vsoch, Imo, I like the variable flag better, just like `ls`, `git` (although git also has a `--no-color` flag).

Comment (vsoch) [reply cmd/internal/cli/singularity.go:109]: Thanks for the feedback! I had totally forgot about this PR. Let me know what changes would be requested, if any.

**Comment by sashayakovtseva** [comment]
@vsoch Can you update your branch, please?

**Review by sashayakovtseva — APPROVED** [review]

**Comment by vsoch** [comment]
I can do the first file, but I don't know how to do the second, it's just been too long. I don't have the branch locally anymore.

**Comment by sashayakovtseva** [comment]
Can't you pull from your remote?

**Comment by vsoch** [comment]
Yes I can, but it produces the same merge conflicts. As I mentioned, I can resolve the first file (singularity.go) but I don't have knowledge about the changes that were made subsequently to sylog_test.go so I wouldn't know how to address the conflicts. The resolution can be done in the interface by someone that has this knowledge. Thanks!

**Comment by sashayakovtseva** [comment]
Ok, I will update this if the rest of the team approves PR.

**Comment by vsoch** [comment]
Great thank you! I gave it a shot, but I cancelled the conflict resolution when I got to the second file. I don't want to mess it up. :P

**Comment by WestleyK** [comment]
Ping @sashayakovtseva

**Comment by sashayakovtseva** [comment]
There are lots of conflicts, honestly.
Looks like creating . a new one on top of the master is easier than to solve them, sorry @vsoch.

**Comment by vsoch** [comment]
yeah, that unfortunately happens when a review takes 6 months for a very busy repository, no worries @sashayakovtseva.

**Pull request closed without merging**
